### ISSUE TITLE
[ENH]: plt.scatter() parameters are extremely confusing

### ISSUE THREAD

**Issue by francescoboc** [matplotlib/matplotlib#27765]
### Problem

Everytime I have to change the marker, size or color of a scatter plot I have to google the parameters because they are impossible to remember. Why is it that:

- to change marker color I can use either the `color` or `c` parameter,
- but to change the marker I can only use `marker`, while the shortened version `m` does not work,
- and to change the marker size only the shortened parameter `s` works but not `size` !!?

So confusing!

![image](https://github.com/matplotlib/matplotlib/assets/33369475/13873a29-85f0-4455-85d8-b118638aafb7)

### Proposed solution

Define parameters to customize the scatter plot more consistently.

**Comment by timhoffm** [comment]
Size `s` and color `c` are the properties you can set per data point `scatter(x, y, s=..., c=...)`. They are used quite often and it's a historic decision that they are single chararcters. Whether we like it or not, that's used too widely to be changed.

`marker` is a single configuration parameter, and written out consistently throught the library (e.g. `plot(..., marker='x')`). It's not reasonable to add a shortcut `m` because that's redundant and not readable.

`color` is a bit special. It's a `Collection` property and we allow setting all such properties via keyword arguments (we basically get that automatically because `scatter` creates a `Collection`. Unfortunately, this clashes with the semantics of the explicitly introduced `c`, so we have to remap `color` to `c`.

The only thing one could debate here is whether one wants to add an additional alias `size` mapping to `s`. But I claim it's not good to have multiple parameter names for the same thing.

I see your issue but I don't see a way forward to make the API more clear while maintaining backward compatibility.

**Comment by francescoboc** [comment]
I understand the motivations, historical and not, that you have provided me for the current situation. But this does not change the fact that the current situation is confusing. Am I supposed to think "oh ok, `s` and `c` are there for historic reasons, but wait `c` can also be set with `color` because `scatter` creates a `Collection` object, while for the markers I have to use `marker` because `m` is not there for historic reasons" every time I need to change properites of a scatter plot? It's impossible to remember!

"it's not reasonable to add a shortcut `m` because that's redundant and not readable." well... also `c` is redundant, and both `c` and `s` are not readable, so I don't really see your point.

The ideal solution for this messy situation would be to add `m` for `marker` ans `size` for `s`, so that all the 3 proprietes are consistently set either with their explicit parameter, or their shortened version. If that is not possible, then yes, at least add `size` for `s`, so that people don't have to remember that marker and color are set with `marker` and `color`, but size is special and only works with `s`...

I am sorry if i sound a little rough, but this problem frustrates me so much, i think i have googled how to set the properties of a scatter plot at least 100 times in my life.

**Comment by rcomer** [comment]
See also #1101.

**Comment by francescoboc** [comment]
> See also #1101.

Indeed! The inconsistency of `ms`/`markersize` and `s` between `plot` and `scatter` also adds confusion to the situation. I didn't want to mention it in the first post because I wanted to focus on the `scatter` function, but that's also a problem that wasted a lot of my time for no reason.

**Comment by jklymak** [comment]
The only reason to use `scatter` over `plot` is to have the marker size and color change on a per-point basis.  If you want all your dots to have the same "size" and "color", then just use `plot`.

**Comment by jklymak** [comment]
> but scatter-> scatter plots and plot-> line plots is a reasonable way to understand/use the API.

Sure, but it would also be misleading.  

The `scatter` API has different names for `s` and `c` because they are meant to be vectors.  If you find yourself saying "why can't I use `markersize` in `scatter`?", rather than saying "Matplotlib should make `scatter` more like `plot`" you should say "this is a situation where using `plot` is more appropriate than `scatter`".

**Comment by francescoboc** [comment]
> > If you want all your dots to have the same "size" and "color", then just use plot.
> 
> We mention this as an optimization, but `scatter`-> scatter plots and `plot`-> line plots is a reasonable way to understand/use the API.

This interpretation is also the one that the majority of new users have (me included).

**Comment by anntzer** [comment]
>>> scatter-> scatter plots and plot-> line plots is a reasonable way to understand/use the API.

>> This interpretation is also the one that the majority of new users have (me included).

> I also think this is not a good interpretation and should not be advertised (@story645 I know you differ here). 

I would go further and say this interpretation is completely wrong (even though it may be common) and we should actively write the docs in a way that goes against this misunderstanding.

**Comment by jklymak** [comment]
@francescoboc there has to be a default, and `plot` happens to default to a line with no marker.  `plt.plot(x, y, 'o')` is the idiomatic way to make just circular markers with no line (in fact `plt.plot(x, y, ls='')` will make an empty plot).  It doesn't mean you _can't_ do `plt.scatter(x, y)`, just that they are two distinct APIs that are provided to do different things, but can be made to overlap.  These idioms are _very_ old (dating to early Matlab, so >40 years), so they are not likely to be changed in any major way.   I also think trying to make the methods overlap more would be an API mistake and lead to even more confusion (for instance `scatter(..., s=10)` and `plot(..., markersize=10)` are _different_ sizes!).  

> we should actively write the docs in a way that goes against this misunderstanding.

For sure, if there are places where we could differentiate better that would be welcome.  

I'll also point out that it is quite easy to write wrappers around our API for your own API (`my_homogenous_scatter(ax, x, y, ...)`).

**Comment by francescoboc** [comment]
> @francescoboc there has to be a default, and `plot` happens to default to a line with no marker. `plt.plot(x, y, 'o')` is the idiomatic way to make just circular markers with no line (in fact `plt.plot(x, y, ls='')` will make an empty plot). It doesn't mean you _can't_ do `plt.scatter(x, y)`, just that they are two distinct APIs that are provided to do different things, but can be made to overlap. These idioms are _very_ old (dating to early Matlab, so >40 years), so they are not likely to be changed in any major way. I also think trying to make the methods overlap more would be an API mistake and lead to even more confusion (for instance `scatter(..., s=10)` and `plot(..., markersize=10)` are _different_ sizes!).
> 
> > we should actively write the docs in a way that goes against this misunderstanding.
> 
> For sure, if there are places where we could differentiate better that would be welcome.
> 
> I'll also point out that it is quite easy to write wrappers around our API for your own API (`my_homogenous_scatter(ax, x, y, ...)`).

Yes, sorry I wrote it quickly and I forgot the marker parameter. What I would normally use is `plt.plot(x, y, ls='', marker='o')`.
I just tried `plt.plot(x, y, 'o')`, and yes it does indeed produce a scatter plot with no lines, thank you for the suggestion! I will use this command from now on instead of `plt.scatter`.

The main confusion for me comes from the fact that, intuitively, if I want to make a simple **scatter** plot (with all points having the same properties) I use **`plt.scatter(x,y)`**, but after this exchange I now understand that the correct function is `plt.plot(x,y,'o')`. This in my opinion is misleading and leads many users to misuse the `scatter` function.

**Comment by anntzer** [comment]
(Perhaps I can take advantage of this discussion to try and revive #14174, by the way?)

**Comment by rcomer** [comment]
It probably doesn't help that `plot` returns an object called `Line2D`, even if no line is drawn.

**Comment by rcomer** [comment]
Would it be worth adding a second entry for `plot` in [plot types](https://matplotlib.org/stable/plot_types/index#pairwise-data) showing a scatter plot?  The `scatter` entry already has random colours and sizes, so it could help illustrate the distinction between what to use for a simple scatter vs a more configured one.

**Comment by timhoffm** [comment]
Friends, the relation between `plot` and `scatter` *is* messy for historic reasons, partly originating even from MATLAB. Just complaining that it's unintuitive or insisting on one certain ill-fitting point of view (plot = line plot) does not make it better for our user. I'd like to have constructive suggestions how to improve the situation.

We only have very limited possibilities to change the API and naming due to backward-compatibility. IMHO we can help the users most by proper description in the documentation. In particular that means *not* primarily associtating `plot` with the visual (line vs. marker) but with the data. The docstring already does this "Plot y versus x as lines and/or markers".

The plot types visual could simply be expanded to ![image](https://github.com/matplotlib/matplotlib/assets/2836374/6474844e-76ba-4fa9-8896-060794ebd520)

**Comment by timhoffm** [comment]
> Would it be worth adding a second entry for `plot` in [plot types](https://matplotlib.org/stable/plot_types/index#pairwise-data) showing a scatter plot?

Cross post :smile: . Yes! See my comment above.

**Comment by rcomer** [comment]
If the thumbnails look like this, it makes it really obvious how much the use-cases can overlap

![plot_scatter](https://github.com/matplotlib/matplotlib/assets/10599679/169f76be-d965-442a-a731-a95ee389753d) ![scatter](https://github.com/matplotlib/matplotlib/assets/10599679/8c251d3b-84e6-4e34-9dad-5634897347df)

**Comment by rcomer** [comment]
> scatter -> every type of scatter except uniform scatter, which this function can also handle just fine.

Just to confuse things further, I recently changed some code from using `scatter` to using `plot` because I wanted [half-filled markers](https://matplotlib.org/stable/gallery/lines_bars_and_markers/marker_reference.html#marker-fill-styles), which I don't think `scatter` can do?

**Comment by jklymak** [comment]
> This in my opinion is misleading and leads many users to misuse the scatter function.

I don't think it's a misuse of `scatter`, but it just has a different API, so needs to be used differently.  `size` and `color` don't work because the `scatter` API allows vectors for `s` and `c`. 

> very common exploratory viz workflow is layering in more encodings so we'd end up encouraging users to start with plot to make the scatter, then move to scatter to encode size and color and so they'd run into more issues w/ parameters not having the same name.

I don't think anyone is encouraging that, necessarily.  If they want to move from `scatter(x, y, s=1)` to `scatter(x, y, s=z)`, that is great and perfectly within the API.  But I'd argue `scatter(x, y, markersize=1)` to `scatter(x, y, s=z)` is worse, and more confusing.  Particularly as `s` and `markersize` are different units: `s=2` is the same as `markersize=sqrt(2)`.  Overloading to `scatter(x, y, markersize=1)` to `scatter(x, y, markersize=z)`, would maybe be OK, but I'm still not a fan as it is a different markersize than for `Line2D` markers.  I think it is more clear to keep it a distinct keyword particular to `scatter`.

**Comment by timhoffm** [comment]
While this is getting quite off-topic, I just want to note that you can have half-filled markers in `scatter`.

```
import matplotlib.pyplot as plt
from matplotlib.markers import MarkerStyle
import numpy as np

x, y = np.random.random((2, 20))
plt.scatter(x, y, s=500, marker=MarkerStyle('o', fillstyle='right'))
x, y = np.random.random((2, 20))
plt.scatter(x, y, s=500, marker=MarkerStyle('o', fillstyle='top'))
```
![image](https://github.com/matplotlib/matplotlib/assets/2836374/8a17a510-aadc-47a4-83a3-47af38e2301b)

So the fundamental mechanism is in place. I suspect however that styling is limited, because the colors and linewidths are not exposed through kwargs.

**Comment by timhoffm** [comment]
A fundamental issue with the marker handling in `scatter()` is that the underlying artist is a `PathCollection` and does not know anything about markers. We convert the marker to a Path in `scatter()`. While one could try and just monkey patch marker kwargs onto scatter, you won’t full functionality: e.g. you still would not have `set_markeredgecolor()`, and also not the property aliases, e.g. `mec`.

I therefore recommend to make a marker-aware subclass `MarkerPathCollection`/`ScatterCollection` if you want to improve marker handling in `scatter()` and gable all logic therein.

Note however, that even then there will remain some rough edges. For example, the rcParams for markers are in the “lines” subgroup.
